lib/web/auth.js

```diff
@@ -14,176 +14,28 @@
  *  limitations under the License.
  *
  */
+var fs = require('fs'),
+    authMethods = {};
 
-var crypto = require('crypto');
-var fs = require('fs');
-
-var AP_MD5PW_ID  = "$apr1$";
-
-var ITOA64 = './0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
-
-var TUPLES = [
-  [0, 6, 12],
-  [1, 7, 13],
-  [2, 8, 14],
-  [3, 9, 15],
-  [4, 10, 5]
-];
-
-
-function AuthDB(users) {
-  this.users = users;
-}
-
-
-AuthDB.prototype.validate = function(username, password, callback) {
-  if (!this.users[username]) {
-    callback(null, false);
-  } else if (this.users[username].indexOf('{SHA}') === 0) {
-    callback(null, this._validateSHA(password, this.users[username]));
-  } else if (this.users[username].indexOf('$apr1$') === 0) {
-    callback(null, this._validateMD5(password, this.users[username]));
-  } else {
-    callback(null, false);
+// read all possible auth handlers (*.js from ./auth folder)
+fs.readdirSync(__dirname + '/auth').forEach(function(authMethod) {
+  if (authMethod.substr(-3) === '.js') {
+    authMethods[authMethod.replace('.js', '')] = require(__dirname + '/auth/' + authMethod);
   }
-};
-
-
-function PassThruAuthDb() {}
+});
 
-PassThruAuthDb.prototype.validate = function(xForwardedUser, password, callback) {
-  if (xForwardedUser) {
-    callback(null, true);
-  } else {
-    callback(null, false);
+exports.initialize = function(config) {
+  // if auth_method is not set in configuration, ensure that the default handler is set to htpasswd or passthru, depending on the login_required config flag
+  if (!config.auth_method) {
+    config.auth_method = (config.login_required ? 'htpasswd' : 'passthru');
   }
-};
-
-
-AuthDB.prototype._validateSHA = function(password, hash) {
-  return sha1crypt(password) === hash;
-};
-
-
-AuthDB.prototype._validateMD5 = function(password, hash) {
-  var all = hash.split('$');
-  return md5crypt(password, all[2], '$apr1$') === hash;
-};
-
-
-AuthDB.prototype.updateUsers = function(users) {
-  this.users = users;
-};
 
-function sha1crypt(password) {
-  return '{SHA}' + crypto.createHash('sha1').update(password).digest('base64');
-}
-
-
-
-// Ported to javascript from http://code.activestate.com/recipes/325204-passwd-file-compatible-1-md5-crypt/
-function md5crypt(password, salt, magic) {
-  var rearranged = '',
-      mixin, final, m, v, i;
-
-  m = crypto.createHash('md5');
-  m.update(password + magic + salt);
-  mixin = crypto.createHash('md5').update(password + salt + password).digest("binary");
-
-  for (i = 0; i < password.length; i++) {
-    m.update(mixin[i % 16]);
+  if (Object.keys(authMethods).indexOf(config.auth_method) < 0) {
```

Comment:
Object.keys(authmethods) is called twice... it makes sense to pull them out into a variable.